src/package_creator.rs

```diff
@@ -0,0 +1,162 @@
+use zip::result::ZipError;
+use zip::write::FileOptions;
+
+use std::fs::{File, create_dir, copy};
+use std::path::Path;
+use walkdir::{DirEntry, WalkDir};
+
+use rocket_contrib::json::Json;
+use serde::{Serialize, Deserialize};
+use std::ops::Add;
+
+#[derive(Deserialize)]
+pub struct Test {
+    input: String,
+    output: String
+}
+
+#[derive(Deserialize)]
+pub struct Task {
+    tag: String,
+    title: String,
+    task_statement: String,
+    exemplary_test: Test,
+    tests: Vec<Test>
+}
+
+// Given task data from form, creates package
+pub fn create(task : Json<Task>) -> zip::result::ZipResult<()> {
+    let tmp_dir = TempDir::new(task.tag.as_str())?;
+    prepare_directories(&tmp_dir, &task)?;
+    prepare_tests(&tmp_dir, &task)?;
+    let res = zip_package(tmp_dir.path().to_str().unwrap(),
+                          &*format!("/tmp/{}.zip", task.tag.as_str()));
+    eprintln!("Finished zip_package result.err() = {:?}", res.err());
+    Ok(())
+}
+
+fn prepare_directories(tmp_dir_path : &TempDir, task : &Json<Task>) -> io::Result<()> {
+    create_dir(tmp_dir_path.path().join("doc"))?;
+    create_dir(tmp_dir_path.path().join("in"))?;
+    create_dir(tmp_dir_path.path().join("out"))?;
+    create_dir(tmp_dir_path.path().join("prog"))?;
+    prepare_makefile(tmp_dir_path, &*task.tag)?;
+    prepare_config(tmp_dir_path, task)?;
+    Ok(())
+}
+
+fn prepare_makefile(tmp_dir_path : &TempDir, task_tag : &str) -> io::Result<()> {
+    let makefile_in_path = tmp_dir_path.path().join("makefile.in");
+    std::fs::write(makefile_in_path.as_path(),
+                   std::fs::read_to_string("./resources/makefile.in.example")
+                       .unwrap()
+                       .replace("tag_placeholder", task_tag))
+}
+
+fn prepare_config(tmp_dir_path : &TempDir, task : &Json<Task>) -> io::Result<()> {
+    let config_path = tmp_dir_path.path().join("config.yml");
+    println!("config_path = {}", config_path.as_path().to_str().unwrap());
+    let tests_config = prepare_tests_config(task.tests.len() as i64);
+    std::fs::write(config_path.as_path(),
+                   std::fs::read_to_string("./resources/config.yml.example")
+                       .unwrap()
+                       .replace("task_title_placeholder", &*task.title)
+                       .add(&*tests_config))
+}
+
+fn prepare_tests_config(tests_num : i64) -> String {
+    let mut res: String = String::from("");
+    let normalized_tests = tests_num / 100;
+    let last_test_points = 100 - ((tests_num - 1) * normalized_tests);
+    for i in 1..tests_num {
+        res = res.add(&*format!("{}: {}\n", i, normalized_tests));
+    }
+    res = res.add(&*format!("{}: {}\n", tests_num, last_test_points));
+    res
+}
+
+fn prepare_tests(tmp_dir_path : &TempDir, task : &Json<Task>) -> io::Result<()> {
+    create_test(tmp_dir_path, 0,
+                &*task.exemplary_test.input, &*task.exemplary_test.output,&*task.tag)?;
+    for i in 1..task.tests.len() + 1 {
```

Comment:
Another way to do this would be `for (i, test) in task.tests.iter().enumerate()`.

Reply:
🧑‍🏭 Done